src/analysis/history.cpp

```diff
@@ -186,4 +186,57 @@ Real ReduceJetMomentumFlux(MeshData<Real> *md) {
 
 } // ReduceJetMomentumFlux
 
+Real ReduceMagneticFluxPhi(MeshData<Real> *md) {
```

Comment:
This pattern is almost identical to the other reduce patterns. I wonder if this can be made generic as something like
```C++
template<typename Function>
Real ReduceHorizonQuantity(MeshData<Real> *md, Function &f) {
  // ...

  Real result = 0.0;
  parthenon::par_reduce(
      parthenon::LoopPatternMDRange(), "Phoebus History for Jet Momentum Flux",
      DevExecSpace(), 0, pack.GetDim(5) - 1, kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i, Real &lresult) {
        const auto &coords = pack.GetCoords(b);
        if (coords.x1f(i) <= xh && xh < coords.x1f(i + 1)) {
          const Real dx1 = coords.Dx(X1DIR, k, j, i);
          const Real dx2 = coords.Dx(X2DIR, k, j, i);
          const Real dx3 = coords.Dx(X3DIR, k, j, i);

          // interp to make sure we're getting the horizon correct
          auto m = (f(pack, geom, pvel_lo, pvel_hi, pb_lo, pb_hi, b, k, j, i + 1) -
                    f(pack, geom, pvel_lo, pvel_hi, pb_lo, pb_hi, b, k, j, i - 1)) /
                   (2.0 * dx1);
          auto flux = (f(pack, geom, pvel_lo, pvel_hi, pb_lo, pb_hi, b, k, j, i) +
                       (xh - coords.x1v(i)) * m) *
                      dx2 * dx3;

          lresult += flux;
        } else {
          lresult += 0.0;
        }
      },
      result);
  return 0.5 * result;
}
```
something like that? Then we could combine the reduction functions for flux with mdot etc. Anything that's a horizon quantity could have a single reduction function? What do you think?

Reply:
So, I tried something very like this: take a look at KHARMA's reductions/reductions.hpp [here](https://github.com/AFD-Illinois/kharma/blob/dev/kharma/reductions/reductions.hpp), which has my best shot at it.

The major difference is, I don't force the reduction body to be a given function type, instead letting it be an arbitrary code block, at the cost of doing some truly unspeakable things with macros.  The flexibility is nice, but it was only really crucial back in the old pre-packing days when I absolutely needed the lambda capture to work.  I'm no longer sure it's a better way to go than the single f(pass, the, whole, world, as, arguments) approach, which I admittedly use in plenty of other places in KHARMA, but there may be some extra use/flexibility I'm not thinking of, and certainly adding new reductions has a lot less boilerplate in the macro version.

Honestly, though, if a function template solution really works, I'd love to help with & adopt something like that in KHARMA.

Reply:
I had been thinking about this too and wasn't sure the cleanest way to do it. I'll play with this in a separate PR